### ISSUE TITLE
The upgrade button is not visible even though a new version of the module is available

### ISSUE THREAD

**Issue by nesrineabdmouleh** [PrestaShop/PrestaShop#32497]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

Upgrade module button is not visible

![Capture d’écran de 2023-05-09 15-43-42](https://github.com/PrestaShop/PrestaShop/assets/24647495/7ae36271-4450-4050-a727-441661bef30a)

### Expected behavior

Upgrade button should be displayed when we have a new version of a module

### Steps to reproduce

1- Go to BO > Module manager page
2- expl : Upload the version 1.0.1 of the module https://github.com/PrestaShop/keycloak_connector_demo/tags 
3- The upgrade button is not visible

### PrestaShop version(s) where the bug happened

8.1.x

### PHP version(s) where the bug happened

8.0

**Comment by florine2623** [comment]
Hello @nesrineabdmouleh ,


I reproduce the issue with PrestaShop version `8.0.x`, `8.1.x` and `develop`, I'll add this to the backlog so it can be fixed.
<img width="1087" alt="Screenshot 2023-05-09 at 17 50 31" src="https://github.com/PrestaShop/PrestaShop/assets/16019289/df152c41-4b8e-49f6-8e7d-13f29f341735">

<img width="1073" alt="Screenshot 2023-05-09 at 17 53 10" src="https://github.com/PrestaShop/PrestaShop/assets/16019289/504ae38e-6f72-4a65-92c7-436d2dce6669">



Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you

**Comment by hibatallahAouadni** [comment]
Duplicate of #28433 which should be fixed by #29512 but it seems this regression is back :sob:

**Comment by RosaBenouamer** [comment]
@paulnoelcholot does this bug exists yet ?

**Comment by paulnoelcholot** [comment]
Yes @RosaBenouamer, it still exists

**Comment by Zoefinders** [comment]
@nesrineabdmouleh what do you mean about _Upload_ , can we have video maybe ? Thank!

**Comment by nesrineabdmouleh** [comment]
> @nesrineabdmouleh what do you mean about _Upload_ , can we have video maybe ? Thank!

https://watch.screencastify.com/v/ZfROR8jZUQPH3myeeOpP

**Comment by jolelievre** [comment]
Hi @nesrineabdmouleh 

I'm not sure there's a bug here, when you upload the module archive the upgrade process is automatically triggered. And you can see after that the version of the module has correctly been updated to 1.0.1 so what is the problem exactly? If the module has already been upgrade automatically we don't need to display the upgrade button

**Comment by nesrineabdmouleh** [comment]
Hello @jolelievre 

The problem is =>  when we install the version 1.0.1 we don't have the upgrade button (the upgrade process is not automatically triggered the version is 1.0.1) and there is a new version 1.0.2 for the module.

**Comment by jolelievre** [comment]
Then something is not clear in the description of your issue you indicate that you uploaded the version 1.0.1 so from what we see in the description and the video everything seems good

Do you mean you uploaded an archive containing version 1.0.2 while the currently installed version is 1.0.1? If so can you please provide the archive used so developers can also use it to reproduce the issue

Thanks in advance

**Comment by matks** [comment]
> Hello @jolelievre
> 
> The problem is => when we install the version 1.0.1 we don't have the upgrade button (the upgrade process is not automatically triggered the version is 1.0.1) and there is a new version 1.0.2 for the module.

@nesrineabdmouleh The module https://github.com/PrestaShop/keycloak_connector_demo is not in the perimeter of the Distribution API.

The Module manager which powers the backoffice page "Module manager" polls the Distribution API for available updates. As keycloak module is not in the perimeter of the distribution API, the distribution API will not tell you there is a newer version available.

This is not a bug, this is the expected behavior.

The backoffice Module Manager page will show an "Upgrade" button
- if the module is in the perimeter of the distribution API https://github.com/prestashop/distribution-api/
- if there is a more recent and published version of the module

**Comment by jolelievre** [comment]
@matks we already discussed this morning but again you focus too much on the keycloak module That's not the point!! The bug would be the same with any module installed manually and the fact that the upload module is not correctly upgraded (wether it's automatically which doesn't seem to be the case, or manually which is impossible because the button is not available)

Actually, it should be reproduced with a custom module dedicated to reproduce this issue and we can move forward on this bug instead of talking about keycloack

**Comment by jolelievre** [comment]
The problem (as explained in another comment) is the same as this old issue https://github.com/PrestaShop/PrestaShop/issues/28433

It was supposed to be fixed by this PR but seems like it was not enough or another regression was introduced back https://github.com/PrestaShop/PrestaShop/pull/29512

**Comment by matks** [comment]
> @matks we already discussed this morning but again you focus too much on the keycloak module That's not the point!! The bug would be the same with any module installed manually and the fact that the upload module is not correctly upgraded (wether it's automatically which doesn't seem to be the case, or manually which is impossible because the button is not available)

I simply comment the video 😋 . If I can see a video with another module, which is in the distribution API perimeter, then I agree it's different.

**Comment by 0x346e3730** [comment]
Reproduced with `GoogleAnalytics 4.2.0` on `develop`.

**Comment by 0x346e3730** [comment]
This happens because `https://api.prestashop-project.org/prestashop` doesn't return `8.1.0` but `8.1.0-beta.1` & `8.1.0-rc.1`so distribution API doesn't work right now for branches `8.1.x` and `develop`, because they both request the endpoint `https://api.prestashop-project.org/modules/<version>` with `<version>` being respectivelly `8.1.0` & `9.0.0`, triggering the folloxing XML responses :

```xml
<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Details>No such object: distribution-api-artifacts-production/modules/8.1.0.json</Details>
</Error>
```

```xml
<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Details>No such object: distribution-api-artifacts-production/modules/9.0.0.json</Details>
</Error>
```

**Comment by matks** [comment]
OK we had a big discussion with @0x346e3730 about this 😁 

It's a problem we already had with 8.0.0 : how to make Distribution API aware of a release not released yet?

I am not confortable with the idea that we would "fake" the release being available in Distribution API _production environment_ `https://api.prestashop-project.org`.

So we are exploring the idea of using the _preproduction environment_ `https://preprod-api.prestashop-project.org`

1. Find out where the JSON https://preprod-api.prestashop-project.org/prestashop is being stored (I know it's on a GCP bucket, but where exactly?)
2. Modify the JSON manually. Indeed the code here https://github.com/PrestaShop/distribution-api/blob/main/src/Command/DownloadNewPrestaShopReleasesCommand.php#L30 makes sure that _the API automatically adds missing GitHub releases but does not remove versions already in the JSON_. So our manual modification should not be wiped out tomorrow when the cronjob that does update the API is triggered.

After step 2 https://preprod-api.prestashop-project.org/prestashop should return 8.1.0 😁 .

Next step would be

3. Submit a PR against https://github.com/PrestaShop/ps_distributionapiclient/ to enable people to either use the production distribution env or the preproduction (or another!). Suggestion from @0x346e3730 : use an environment variable, it's docker user friendly.

If the 3 steps can be achieved, then _everytime in the future we need a usecase where ps_distributionapiclient need to be fed data about an un-published release_ we can use the preproduction environment.

We could do this for both 8.1.0 and 9.0.0 .

**Comment by 0x346e3730** [comment]
The PRs I've sent will show a warning when this happens on the modules list now, however we need to create another issue to make Distribution API work with unreleased versions as discussed in the above comments.
